### ISSUE TITLE
Extend global detection

### ISSUE THREAD

**Issue by guybedford** [systemjs/builder#211]
We need to detect `window.x`, `self.x` and `global.x` as globals in the cases where `window`, `self` and `global` are not otherwise declared in scope as part of the global compiler global variable detection.

**Comment by OliverJAsh** [comment]
I'm not sure I understand the fix you are describing. Could you elaborate? I don't know much about global detection as it stands today. I would like to understand how the code I provided in my example will work with this change.

**Comment by rich-nguyen** [comment]
Thanks @guybedford , do you mean extending detection for variables that are implicitly defined as part of `window`, `self`, and `global`?

**Comment by guybedford** [comment]
Yes exactly, this is just about detecting `window.jQuery = {...}` in the ast to know that `jQuery` is a global. We currently only handle variable declarations in https://github.com/systemjs/builder/blob/master/compilers/global.js#L28, so just need to add the above adjustments to the transformer. It should be relatively straightforward.

**Comment by OliverJAsh** [comment]
So `function foo() {}` would become `window.foo = function () {}`?

**Comment by guybedford** [comment]
We add the global assignments at the end in https://github.com/systemjs/builder/blob/master/compilers/global.js#L70.

Thanks, `function globalName() {}` is a case we also need to cover though too.

**Comment by OliverJAsh** [comment]
Thanks, global function declarations was the issue I reported in https://github.com/systemjs/systemjs/issues/545.

**Comment by guybedford** [comment]
Ah, right, I misread that. Ok that case is much easier - included in https://github.com/systemjs/builder/commit/1c8afd53785db45ccc3f65f2965e6047936219df.

**Comment by OliverJAsh** [comment]
Great, thank you. I'm not sure how this will actually transform input code, i.e.

``` js
function foo() {}
window.foo();
```

This currently bundles as:

``` js
(function () {
    function foo() {}
    window.foo();
});
```

Which obviously breaks the code, because the function declaration is not on the root scope, and thus is not added to `window`.

Will the change you introduced make the above bundle to the following?

``` js
(function () {
    function foo() {}
    // Force function declaration onto outer scope:
    window.foo = foo;
    // So that this will work:
    window.foo();
});
```

Or something different?

**Comment by guybedford** [comment]
Sure!

The output of the file at https://github.com/systemjs/builder/blob/master/test/fixtures/test-tree/global-outer.js comes out as:

```javascript
System.registerDynamic("global-outer.js", [], false, function(__require, __exports, __module) {
  var _retrieveGlobal = System.get("@@global-helpers").prepareGlobal(__module.id, null, null);
  (function() {
    var p = this["p"];
    var q = this["q"];
    var window = this["window"];
    (function() {
      p = 6;
    });
    var p = 5;
    function q() {}
    if (false)
      var window = 5;
    this["p"] = p;
    this["q"] = q;
    this["window"] = window;
  })();
  return _retrieveGlobal();
});

```

**Comment by guybedford** [comment]
I guess the duplicate vars are not ideal here, and should be filtered, but it works correctly as valid ES3.

**Comment by OliverJAsh** [comment]
I've just had a go at testing this. I think we need to hoist the assignment, otherwise code like this won't work:

``` js
// The function declaration is hoisted, so this works
window.foo();
function foo () {}
```

Currently we convert this to:
``` js
this['foo']();
function foo () {}
this['foo'] = foo;
```

Whereas it needs to be:

``` js
this['foo'] = foo;
this['foo']();
function foo () {}
```

**Comment by guybedford** [comment]
Thanks - ok yes we need to cover the hoisting. I'll take another stab at this with a hoisting approach.

Also to update on the original issue - we obviously do support assignment to the global object, but do so via the helper function wrappings which do the window diff - this way we have a method that always works in the browser and with builds equivalently.

So just handling this function case is what needs to be done here. Will get to it as soon as I have a moment.

**Comment by guybedford** [comment]
I believe I've fixed this in the 0.12.2 release. Please let me know if that solves your issue.

**Comment by OliverJAsh** [comment]
This solves our issue. Thank you very much!

**Comment by guybedford** [comment]
Released in 0.18.5, sorry for the long delay on this one, was just bad timing unfortunately.
